File: client/src/main/java/dev/thompgt/habitsync/client/SyncWire.java
package dev.thompgt.habitsync.client;

import dev.thompgt.habitsync.sync.WireChange;
import java.util.List;
import java.util.UUID;

/**
 * The JSON bodies of {@code /v1/sync}, as this client sends and expects them.
 *
 * <p>A separate declaration from the server's {@code SyncDtos} rather than a shared one, and
 * that is a real trade-off rather than an oversight. Sharing would guarantee the two agree;
 * it would also drag the client's dependencies onto whatever the server's DTO module happens
 * to need, and would make a protocol change look source-compatible when it is not. What is
 * shared is the part where drift actually costs data — {@link WireChange} and its codec live
 * in {@code sync-core} and both ends encode through them.
 *
 * <p>The envelope around it is small enough to be checked by an end-to-end test, and there is
 * one: if a field name here stops matching the server's, that test fails rather than a device
 * silently syncing nothing.
 */
final class SyncWire {

    private SyncWire() {}

    /** Must match the server's {@code SyncDtos.PROTOCOL_VERSION}, or it answers 426. */
    static final int PROTOCOL_VERSION = 1;

    record Request(long sinceSeq, int protocolVersion, List<WireChange> ops) {}

    record Response(
            List<UUID> appliedOpIds,
            List<ChangeEnvelope> changes,
            long nextSeq,
            boolean hasMore,
            boolean resyncRequired,
            String resyncReason,
            long serverTimeMillis,
            int protocolVersion) {}

    record ChangeEnvelope(long serverSeq, WireChange change) {}

    record LoginRequest(String email, String password, String deviceName, UUID deviceId) {}

    record RegisterRequest(String email, String password, String deviceName) {}

    record RefreshRequest(String refreshToken) {}

    record TokenResponse(
            String accessToken,
            String refreshToken,
            String tokenType,
            long expiresIn,
            UUID userId,
            UUID deviceId) {

        Session toSession() {
            return new Session(userId, deviceId, accessToken, refreshToken);
        }
    }
}

File: ios/Sources/HabitSyncClient/SyncWire.swift
import Foundation
import HabitSyncCore

/// The JSON bodies of `/v1/sync` and `/v1/auth`, as this client sends and expects them.
///
/// A separate declaration from the server's `SyncDtos` rather than a shared one, and that is
/// a real trade-off rather than an oversight — the two are in different languages here, so
/// sharing was never on the table, but the same discipline applies: what is shared is the
/// part where drift actually costs data. ``WireChange`` and ``ChangeCodec`` are ported
/// field-for-field from `sync-core`, and both ends encode changes through them. The envelope
/// around them is small enough to be checked by an end-to-end test against a running server.
enum SyncWire {

    /// Must match the server's `SyncDtos.PROTOCOL_VERSION`, or it answers 426.
    static let protocolVersion = 1

    struct Request: Encodable {
        let sinceSeq: Int64
        let protocolVersion: Int
        let ops: [WireChange]
    }

    struct Response: Decodable {
        // Optional on the way in because a server is free to omit an empty list, and a
        // decoding failure here would look to the engine like an unsyncable device.
        let appliedOpIds: [UUID]?
        let changes: [ChangeEnvelope]?
        let nextSeq: Int64
        let hasMore: Bool
        let resyncRequired: Bool
        let resyncReason: String?
        let serverTimeMillis: Int64
        let protocolVersion: Int
    }

    struct ChangeEnvelope: Decodable {
        let serverSeq: Int64
        let change: WireChange
    }

    struct LoginRequest: Encodable {
        let email: String
        let password: String
        let deviceName: String
        /// The caller's existing device id, if it has one. Supplying it keeps the device's
        /// HLC node identity stable across sign-ins; omitting it registers a new device.
        let deviceId: UUID?
    }

    struct RegisterRequest: Encodable {
        let email: String
        let password: String
        let deviceName: String
    }

    struct RefreshRequest: Encodable {
        let refreshToken: String
    }

    struct TokenResponse: Decodable {
        let accessToken: String
        let refreshToken: String
        let tokenType: String?
        let expiresIn: Int64
        let userId: UUID
        let deviceId: UUID
    }
}
